### PR TITLE
tezos-node: init at 11.1

### PR DESCRIPTION
**Pull request by ulrikstrid**
###### Motivation for this change

This adds `tezos-node` as both a binary and service to run it in NixOS

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/blockchains/tezos/node.nix [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+, makeWrapper
+, zcash
+}:
+
+ocamlPackages.buildDunePackage rec {
+  pname = "tezos-node";
+  inherit (ocamlPackages.tezos-stdlib) version useDune2;
+  src = "${ocamlPackages.tezos-stdlib.base_src}/src/bin_node";
+
+  propagatedBuildInputs = [zcash];
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/tezos/node.nix:16]:
```suggestion
  propagatedBuildInputs = [ zcash ];
```

File: pkgs/development/ocaml-modules/tezos/protocol-make.nix [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+rec {
+  protocol = buildDunePackage {
+    pname = "tezos-protocol-${protocol-name}";
+    inherit (tezos-stdlib) version useDune2;
+    src = "${src}/lib_protocol";
+
+    preBuild = ''
+      rm dune
+      cp -f ${tezos-protocol-compiler.src}/dune_protocol dune
+      sed -i.back -e s/-nostdlib//g dune.inc
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/tezos/protocol-make.nix:22]: Please use substituteInPlace and we also don't need a backup.

File: nixos/modules/services/blockchain/tezos/node.nix [hunk 3]
```diff
@@ -0,0 +1,152 @@
+# SPDX-FileCopyrightText: 2021 TQ Tezos <https://tqtezos.com/>
+#
+# SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+{config, lib, pkgs, ...}:
```
Comment (SuperSandro2000) [inline nixos/modules/services/blockchain/tezos/node.nix:5]:
```suggestion
{ config, lib, pkgs, ... }:
```

File: nixos/modules/services/blockchain/tezos/node.nix [hunk 4]
```diff
@@ -0,0 +1,152 @@
+# SPDX-FileCopyrightText: 2021 TQ Tezos <https://tqtezos.com/>
+#
+# SPDX-License-Identifier: LicenseRef-MIT-TQ
+
```
Comment (SuperSandro2000) [inline nixos/modules/services/blockchain/tezos/node.nix:4]:
```suggestion
```

All files in this repo are licenses under MIT.

File: nixos/modules/services/blockchain/tezos/common.nix [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    serviceConfig = {
+      User = "tezos-${node-name}";
+      Group = "tezos-${node-name}";
+      StateDirectory = "tezos-${node-name}";
+      Restart = "always";
+      RestartSec = "10";
+    };
+  };
+
+}
```
Comment (SuperSandro2000) [inline nixos/modules/services/blockchain/tezos/common.nix:98]:
```suggestion
  };
}
```

File: nixos/modules/services/blockchain/tezos/common.nix [hunk 6]
```diff
@@ -0,0 +1,98 @@
+# SPDX-FileCopyrightText: 2021 TQ Tezos <https://tqtezos.com/>
+#
+# SPDX-License-Identifier: LicenseRef-MIT-TQ
+
```
Comment (SuperSandro2000) [inline nixos/modules/services/blockchain/tezos/common.nix:4]:
```suggestion
```

Comment (ulrikstrid) [reply nixos/modules/services/blockchain/tezos/common.nix:4]: This whole file is copied from Serokell, do I need their permission to remove the license header? cc @balsoft

Comment (kirelagin) [reply nixos/modules/services/blockchain/tezos/common.nix:4]: The terms of [the license](https://github.com/serokell/tezos-packaging/blob/master/LICENSES/LicenseRef-MIT-TQ.txt) require you to do two things: when copying the code, include the original copyright notice and the permission notice. The nixpkgs repository is licensed under MIT, so the “permissions notice” part is covered, however you still do need to include the copyright notice. Frankly, the easiest way to comply with the license would be to keep the [REUSE](https://reuse.software/) header in the file – that’s what it is there for.

If you want, you can move the copyright notice elsewhere, but it has to be in the nixpkgs repository (preferably, in some way that makes it clear that it applies only to these files). If you want to remove the notice entirely, then you need permission not from Serokell, but from the copyright holder, which is TQ in this case.

Comment (ulrikstrid) [reply nixos/modules/services/blockchain/tezos/common.nix:4]: So, is it OK to just keep these headers that I already included? I'm clearly not a lawyer 😅

**Comment by ulrikstrid** [comment]
@SuperSandro2000 I have now removed the problematic files from this PR so that we just add the tezos-node and some refactors in the underlying ocamlPackages to make it easier to maintain in the future.

**Comment by ulrikstrid** [comment]
Since the changes are mainly in ocamlPackages, would you be available for a look @vbgl and/or @symphorien?

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -30444,6 +30444,8 @@ with pkgs;
 
   tessera = callPackage ../applications/blockchains/tessera { };
 
+  tezos-node = callPackage ../applications/blockchains/tezos/node.nix { ocamlPackages =  ocaml-ng.ocamlPackages_4_12; };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:30447]:
```suggestion
  tezos-node = callPackage ../applications/blockchains/tezos/node.nix { ocamlPackages = ocaml-ng.ocamlPackages_4_12; };
```

File: pkgs/applications/blockchains/tezos/node.nix [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    lwt-exit
+    tls
+    cstruct
+  ];
+
+  checkInputs = with ocamlPackages; [
+    tezos-base-test-helpers
+  ];
+
+  postInstall =''
```
Comment (SuperSandro2000) [inline pkgs/applications/blockchains/tezos/node.nix:62]:
```suggestion
  postInstall = ''
```

**Pull request closed without merging**